lib/posix-process/execve.c

```diff
@@ -0,0 +1,198 @@
+
+void execve_arch_execenv_init(struct ukarch_execenv *execenv_new,
+			      struct ukarch_execenv *execenv,
+			      __uptr ip, __uptr sp);
+
+UK_EVENT(POSIX_PROCESS_EXECVE_EVENT);
+
+/* arg0: execenv_new, arg1: stack_old */
+static void __noreturn execve_ctx_switch(long arg0, long arg1)
+{
+	struct ukarch_execenv *execenv_new;
+	struct posix_thread *pthread_parent;
+	struct posix_thread *pthread;
+	struct uk_thread *this_thread;
+	void *stack_old;
+
+	execenv_new = (struct ukarch_execenv *)arg0;
+	stack_old = (void *)arg1;
+
+	this_thread = uk_thread_current();
+	UK_ASSERT(this_thread);
+
+	pthread = tid2pthread(ukthread2tid(this_thread)); /* FIXME */
+	UK_ASSERT(pthread);
+
+	pthread_parent = pthread->parent;
+
+	/* If there is no parent, then must be calling execve()
+	 * from init without having called vfork().
+	 */
+	if (!pthread_parent) {
+		UK_ASSERT(pthread->process);
+		UK_ASSERT(pthread->process->pid == 0);
+		UK_ASSERT(pthread_parent->state != POSIX_THREAD_BLOCKED_VFORK);
+		goto switch_ctx;
+	}
+
+	/* If we're coming from VFORK it's time to wake up the parent.
+	 * Otherwise, free the old stack.
+	 */
+	if (pthread_parent->state == POSIX_THREAD_BLOCKED_VFORK) {
+		uk_pr_debug("Waking up parent (tid %d)\n", pthread_parent->tid);
+		uk_thread_wake(tid2ukthread(pthread_parent->tid));
+		pthread_parent->state = POSIX_THREAD_RUNNING;
+	} else {
+		uk_free(this_thread->_mem.stack_a, stack_old);
+	}
+
+	uk_pr_debug("Switching context\n");
+switch_ctx:
+	ukarch_execenv_load((long)execenv_new);
+}
+
+/* Prepare process for executing new context. For a complete list see
+ * "Effect on process attributes" in execve(2).
+ */
+static int pprocess_cleanup(struct uk_thread *thread __maybe_unused)
+{
+	/* Kill this thread's siblings */
+	pprocess_kill_siblings(thread);
+
+	return 0;
+}
+
+UK_LLSYSCALL_R_E_DEFINE(int, execve, const char *, pathname,
+			char *const *, argv,
+			char *const *, envp)
+{
+	struct posix_process_execve_event_data event_data;
+	struct uk_binfmt_loader_args loader_args;
+	struct ukarch_execenv *execenv_new;
+	struct uk_thread *this_thread;
+	struct ukarch_ctx ctx_old;
+	void *stack_old;
+	void *stack_new;
+	int rc = 0;
+
+	/* Linux deviates from POSIX by treating NULL pointers to
+	 * argv / envp equivalently to passing a list with a single
+	 * NULL element. Preserve this behavior for compatibiltiy
+	 * and treat NULL pointers as valid.
+	 */
+	UK_ASSERT(pathname);
+
+	this_thread = uk_thread_current();
+	UK_ASSERT(this_thread);
+
+	loader_args.argv = (const char **)argv;
+	loader_args.envp = (const char **)envp;
+	loader_args.pathname = (char *)pathname;
+	loader_args.loader = NULL;
+	loader_args.user = NULL;
+
+	/* Assume that if argv is set the caller follows the convention */
+	loader_args.progname = (argv && argv[0]) ? argv[0] : "<null>";
+
+	/* Assign the default allocator to the loader. This will be used
+	 * to allocate memory for the executable image.
+	 */
+	loader_args.alloc = uk_alloc_get_default();
```

Comment:
I am wondering... should we maybe use current thread's allocator instead? In practice for now this will be the same thing sure but maybe in the future we will have a use case for separate per-thread allocators.

Reply:
This is intentional. Notice that below I use the thread's allocator for the new stack. The image uses the default allocator similarly to what we do in the `app-elfloader`, so I see it more as a matter of policy. It also avoids the potential issue of the image occupying a significant amount of the thread's dynamic memory pool.